rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,977 @@
+      this.pendingCall = pendingCall;
+      this.backoffPolicy = backoffPolicy == null ? backoffProvider.get() : backoffPolicy;
+      pendingCall.addListener(
+          new Runnable() {
+            @Override
+            public void run() {
+              handleDoneFuture();
+            }
+          },
+          synchronizationContext);
+    }
+
+    private void handleDoneFuture() {
+      synchronized (lock) {
+        pendingCallCache.remove(request);
+        if (pendingCall.isCancelled()) {
+          return;
+        }
+
+        try {
+          transitionToDataEntry(pendingCall.get());
+        } catch (Exception e) {
+          if (e instanceof ThrottledException) {
+            transitionToBackOff(Status.RESOURCE_EXHAUSTED.withCause(e));
+          } else {
+            transitionToBackOff(Status.fromThrowable(e));
+          }
+        }
+      }
+    }
+
+    private void transitionToDataEntry(RouteLookupResponse routeLookupResponse) {
+      synchronized (lock) {
+        linkedHashLruCache.cache(request, new DataCacheEntry(request, routeLookupResponse));
+      }
+    }
+
+    private void transitionToBackOff(Status status) {
+      synchronized (lock) {
+        linkedHashLruCache.cache(request, new BackoffCacheEntry(request, status, backoffPolicy));
+      }
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("request", request)
+          .add("pendingCall", pendingCall)
+          .add("backoffPolicy", backoffPolicy)
+          .toString();
+    }
+  }
+
+  /** Common cache entry data for {@link RlsAsyncLruCache}. */
+  abstract class CacheEntry {
+
+    protected final RouteLookupRequest request;
+
+    CacheEntry(RouteLookupRequest request) {
+      this.request = checkNotNull(request, "request");
+    }
+
+    abstract int getSizeBytes();
+
+    final boolean isExpired() {
+      return isExpired(timeProvider.currentTimeNanos());
+    }
+
+    abstract boolean isExpired(long now);
+
+    abstract void cleanup();
+  }
+
+  /** Implementation of {@link CacheEntry} contains valid data. */
+  private final class DataCacheEntry extends CacheEntry {
+    private final RouteLookupResponse response;
+    private final long expireTime;
+    private final long staleTime;
+    private ChildPolicyWrapper childPolicyWrapper;
+
+    DataCacheEntry(RouteLookupRequest request, final RouteLookupResponse response) {
+      super(request);
+      this.response = checkNotNull(response, "response");
+      childPolicyWrapper =
+          refCountedChildPolicyWrapperFactory
+              .createOrGet(response.getTarget());
+      long now = timeProvider.currentTimeNanos();
+      expireTime = now + maxAgeNanos;
+      staleTime = now + staleAgeNanos;
+
+      if (childPolicyWrapper.getPicker() != null) {
+        // using cached childPolicyWrapper
+        updateLbState();
+      } else {
+        createChildLbPolicy();
+        linkedHashLruCache.updateEntrySize(request);
```

Comment:
Why `updateEntrySize` in constructor? It looks wired the constructor can modify linkedHashLruCache's state.

Reply:
yes it should be called outside. but for current impl it is constant, so removed it.